### PR TITLE
[RFC] Reduce nf4 quant mem

### PR DESCRIPTION
**Pull request by ebsmothers**
Currently NF4 quantization of large tensors can cause some pretty substantial memory spikes. Consider the following script for quantizing the Llama3-8B output projection weight into NF4:

```
from torch import nn
from torchtune.utils import get_memory_stats, get_device
from torchao.dtypes.nf4tensor import to_nf4

def main():
    
    device = get_device('cuda')

    # Size of Llama3-8B output projection weight
    big_linear = nn.Linear(in_features=4096, out_features=128256, bias=False, device=device)
    memory_stats = get_memory_stats(device=device)
    print(f"before quantize: {memory_stats}")
    
    # Quantize with ao
    ao_quant = to_nf4(big_linear.weight)
    memory_stats = get_memory_stats(device=device)
    print(f"after ao quant: {memory_stats}")



if __name__ == "__main__":
    main()
```

On main, this currently prints
```
before quantize: {'peak_memory_active': 2.101346304, 'peak_memory_alloc': 2.101346304, 'peak_memory_reserved': 2.101346304}
after ao quant: {'peak_memory_active': 37.865276416, 'peak_memory_alloc': 37.865276416, 'peak_memory_reserved': 37.977325568}
```

The peak memory is almost 38 GB, I think due to the fact that `diff = (value - nf4).abs()` is creating a 16x larger tensor to do the pairwise comparison of `value` to `nf4` tensors. 

**Edit**: Updated based on @gau-nernst and @cpuhrsch's suggestions. Instead of a for loop we just `torch.compile` `quantize_tensor_nearest` (conditional upon not running on windows) then cast the final indices to uint8. This results in

```
before quantize: {'peak_memory_active': 2.101346304, 'peak_memory_alloc': 2.101346304, 'peak_memory_reserved': 2.101346304}
after ao quant: {'peak_memory_active': 5.29545728, 'peak_memory_alloc': 5.29545728, 'peak_memory_reserved': 5.668601856}
```

### CODE REVIEW

**Comment by gau-nernst** [comment]
Can't we just `torch.compile()` the function `quantize_tensor_nearest()`?

**Comment by ebsmothers** [comment]
@gau-nernst thanks for the suggestion. I decorated `quantize_tensor_nearest` with `@torch.compile` and still see a peak allocated memory around 12.7GB. Any ideas on how I could reduce it further?

**Comment by gau-nernst** [comment]
On my phone right now so I can't check in detail. From the look of it, I think the original implementation return int32 indices, while your updated version return uint8. That might account for the difference in memory usage. Try .to(torch.uint8) in the torch.compile version also. I think torch.compile should be able to generate efficient triton code for this (can check with TORCH_LOGS="output_code"

**Comment by ebsmothers** [comment]
@gau-nernst that gets it to comparable memory, thanks! Is it reasonable to change the PR to just add the uint8 cast + torch.compile decorator to this method then?

**Comment by drisspg** [comment]
I think this makes sense, the other way I would do this, which would maybe be more performant, is to do this step in chunks similar to this PR:

https://github.com/pytorch/ao/pull/196

**Comment by cpuhrsch** [comment]
@ebsmothers - I'd pick the one that sufficiently reduces memory without introducing performance regressions. So if you can get away with just using compile and not doing chunking, maybe that'll preserve perf and also fix your memory issue.

**Comment by gau-nernst** [comment]
Just a comment. We can probably torch.compile `convert_to_norm_float_weight()` also, which further reduce memory usage (if users don't call `quantize_tensor_nearest()` directly). (and chunking logic can be removed, unless you are on Windows 😆)

**Comment by drisspg** [comment]
IMO a big part of this subclass is that it's an eager implementation is very easy to hack on and easy to debug. I am in general not a big fan of baking in torch.compile

**Comment by cpuhrsch** [comment]
@drisspg - Yes, that's a good point. If we turn it on torch.compile by default, it'll be harder to debug.

@gau-nernst - Maybe we'll just add a test that verifies that torch.compile works. So we'll write the code in a way that it's compile-able and results in efficient code, but the user does still have to turn on compile themselves. What do you think?

**Comment by cpuhrsch** [comment]
@gau-nernst - Hm, generally `torch.compile` *should* work on most PyTorch code, albeit it might cause graph breaks (they can be detected with `TORCH_LOGS='graph_breaks'`), which could impact performance. If you tried to compile something and it didn't work, I think it'd be worth for us to see if we can fix it. In any case, it's also ok to have multiple calls to `torch.compile` within a process if there's code that prevents us from using it (but really if there is such code, we should likely figure out why for best overall performance).

**Comment by gau-nernst** [comment]
@cpuhrsch ok, I will look through some of our public quant API and open a separate issue if they can't run through torch.compile.

File: torchao/dtypes/nf4tensor.py [hunk 1]
```diff
@@ -698,7 +699,7 @@ def quantize_tensor_nearest(
         value = value.unsqueeze(-1)  # (numel, 1)
         # Compare the value tensor with the nf4 tensor element-wise
         diff = (value - nf4).abs()
-        closest_nf4 = diff.min(dim=-1).indices
+        closest_nf4 = diff.min(dim=-1).indices.to(dtype=torch.uint8)
```
Comment (cpuhrsch) [inline torchao/dtypes/nf4tensor.py:702]: On another note, it seems like we could use 4 bits here? We have https://github.com/pytorch/ao/blob/03e2c9b056a17a0e6d5ae31c2a73df36345f886f/torchao/dtypes/uint4.py , but it likely needs some work. I assume `closest_nf4` here is an index into the nf4 table which by definition only is 4 bits and thus there aren't more than 4bit worth of indices?

Comment (gau-nernst) [reply torchao/dtypes/nf4tensor.py:702]: Bit-packing logic is done in `convert_to_norm_float_weight()`. I don't know the original author's intention but one benefit of not doing bit-packing in `quantize_tensor_nearest()` is that we can do correctness test (without unpacking bits).

Comment (cpuhrsch) [reply torchao/dtypes/nf4tensor.py:702]: Mhm, I see. So then using compile at a higher level might still avoid materializing the full int8 tensor (thus reducing peak memory utilization during construction).

Comment (drisspg) [reply torchao/dtypes/nf4tensor.py:702]: Also, from what I remember when I wrote this, the bitshifting is only supported on uint8 dtype: https://github.com/pytorch/ao/blob/03e2c9b056a17a0e6d5ae31c2a73df36345f886f/torchao/dtypes/nf4tensor.py#L661

**Comment by drisspg** [comment]
Another TBH is that the pragmatist in me is just use this

**Comment by drisspg** [comment]
Also one other thing I noticed is that I think you are using an older version of torch ao that doesn't have my chunking PR:

I get this:

```
before quantize: {'peak_memory_active': 2.101346304, 'peak_memory_alloc': 2.101346304, 'peak_memory_reserved': 2.101346304}
after ao quant: {'peak_memory_active': 5.295585792, 'peak_memory_alloc': 5.295585792, 'peak_memory_reserved': 5.49453824}
```

**Comment by drisspg** [comment]
So I still kind of think all the memory savings is just coming from the previous chunking pr?

File: torchao/dtypes/nf4tensor.py (hunk 1 above)
Comment (vadimkantorov) [inline torchao/dtypes/nf4tensor.py:702]: Regarding, `diff.min(dim=-1).indices.to(torch.uint8)`: if `diff.argmin(dim = -1, dtype = torch.uint8)` [is supported](https://github.com/pytorch/pytorch/issues/32570), this might get neater

**Comment by ebsmothers** [comment]
> So I still kind of think all the memory savings is just coming from the previous chunking pr?

o god.. thank you @drisspg, this is already handled by the PR you landed previously. My torchtune env with ao==0.1.0 was evidently bleeding into my dev env and I did not catch it. I agree your existing PR solves this problem exactly, sorry for the thrash here. I am gonna close this and get torchtune on 0.2.0 asap
